File: dvbViewerController/src/main/java/org/dvbviewer/controller/ui/fragments/TimerList.java
/*
 * Copyright © 2013 dvbviewer-controller Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.dvbviewer.controller.ui.fragments;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.LoaderManager.LoaderCallbacks;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.view.ActionMode.Callback;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import com.nostra13.universalimageloader.utils.IoUtils;

import org.dvbviewer.controller.R;
import org.dvbviewer.controller.entities.Timer;
import org.dvbviewer.controller.io.AuthenticationException;
import org.dvbviewer.controller.io.DefaultHttpException;
import org.dvbviewer.controller.io.ServerRequest;
import org.dvbviewer.controller.io.data.TimerHandler;
import org.dvbviewer.controller.ui.base.AsyncLoader;
import org.dvbviewer.controller.ui.base.BaseActivity.AsyncCallback;
import org.dvbviewer.controller.ui.base.BaseListFragment;
import org.dvbviewer.controller.ui.phone.TimerDetailsActivity;
import org.dvbviewer.controller.ui.widget.ClickableRelativeLayout;
import org.dvbviewer.controller.utils.ArrayListAdapter;
import org.dvbviewer.controller.utils.DateUtils;
import org.dvbviewer.controller.utils.ServerConsts;
import org.dvbviewer.controller.utils.UIUtils;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


/**
 * The Class TimerList.
 *
 * @author RayBa
 * @date 07.04.2013
 */
public class TimerList extends BaseListFragment implements AsyncCallback, LoaderCallbacks<List<Timer>>, Callback, OnClickListener, TimerDetails.OnTimerEditedListener, DialogInterface.OnDismissListener, AdapterView.OnItemLongClickListener {

	public static final String ACTION_MODE        = "action_mode";
	public static final String CHECKED_ITEM_COUNT = "checked_item_count";
	TimerAdapter	mAdapter;
	ActionMode		mode;
	ProgressDialog	progressDialog;
	private boolean actionMode;

	/* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onCreate(android.os.Bundle)
	 */
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		mAdapter = new TimerAdapter(getActivity());
		setHasOptionsMenu(true);
	}

	/* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onActivityCreated(android.os.Bundle)
	 */
	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		setListAdapter(mAdapter);
		Loader<List<Timer>> loader = getLoaderManager().initLoader(0, savedInstanceState, this);
		setListShown(!(!isResumed() || loader.isStarted()));
		getListView().setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
		getListView().setOnItemLongClickListener(this);
		setEmptyText(getResources().getString(R.string.no_timer));
		if (savedInstanceState != null && savedInstanceState.getBoolean(ACTION_MODE, false)) {
			AppCompatActivity activity = (AppCompatActivity) getActivity();
			mode = activity.startSupportActionMode(this);
			updateActionModeTitle(savedInstanceState.getInt(CHECKED_ITEM_COUNT));
		}else{
			getActivity().setTitle(R.string.timer);
		}
	}

	/* (non-Javadoc)
	 * @see android.support.v4.app.LoaderManager.LoaderCallbacks#onCreateLoader(int, android.os.Bundle)
	 */
	@Override
	public Loader<List<Timer>> onCreateLoader(int arg0, Bundle arg1) {
		AsyncLoader<List<Timer>> loader = new AsyncLoader<List<Timer>>(getContext()) {

			@Override
			public List<Timer> loadInBackground() {
				List<Timer> result = null;
				InputStream xml = null;
				try {
					xml = ServerRequest.getInputStream(ServerConsts.REC_SERVICE_URL + ServerConsts.URL_TIMER_LIST);
					TimerHandler hanler = new TimerHandler();
					result = hanler.parse(xml);
					Collections.sort(result);
				} catch (Exception e) {
					catchException(getClass().getSimpleName(), e);
				}finally {
					IoUtils.closeSilently(xml);
				}
				return result;
			}
		};
		return loader;
	}

	/* (non-Javadoc)
	 * @see android.support.v4.app.LoaderManager.LoaderCallbacks#onLoadFinished(android.support.v4.content.Loader, java.lang.Object)
	 */
	@Override
	public void onLoadFinished(Loader<List<Timer>> arg0, List<Timer> arg1) {
		mAdapter.setItems(arg1);
		setListShown(true);
	}

	/* (non-Javadoc)
	 * @see android.support.v4.app.LoaderManager.LoaderCallbacks#onLoaderReset(android.support.v4.content.Loader)
	 */
	@Override
	public void onLoaderReset(Loader<List<Timer>> arg0) {
        arg0.reset();
		if (isVisible()) {
			setListShown(true);
		}
	}

    @Override
    public void timerEdited(boolean edited) {
        if(edited){
            timerSavedAction();
        }
    }

	@Override
	public void onDismiss(DialogInterface dialog) {
		getActivity().setTitle(R.string.timer);
	}

	/**
	 * The Class ViewHolder.
	 *
	 * @author RayBa
	 * @date 07.04.2013
	 */
	private static class ViewHolder {
		ClickableRelativeLayout	layout;
		ImageView				recIndicator;
		TextView				title;
		TextView				channelName;
		TextView				date;
	}

	/**
	 * The Class TimerAdapter.
	 *
	 * @author RayBa
	 * @date 07.04.2013
	 */
	public class TimerAdapter extends ArrayListAdapter<Timer> {


		/**
		 * The Constructor.
		 *
		 * @param context the context
		 * @author RayBa
		 * @date 04.06.2010
		 * @description Instantiates a new recording adapter.
		 */
		public TimerAdapter(Context context) {
			super();
		}

		/*
		 * (non-Javadoc)
		 * 
		 * @see android.widget.ArrayAdapter#getView(int, android.view.View,
		 * android.view.ViewGroup)
		 */
		@Override
		public View getView(int position, View convertView, ViewGroup parent) {
			ViewHolder holder;
			if (convertView == null) {
				convertView = LayoutInflater.from(getContext()).inflate(R.layout.list_item_timer, parent, false);
				holder = new ViewHolder();
				holder.layout = (ClickableRelativeLayout) convertView;
				holder.recIndicator = (ImageView) convertView.findViewById(R.id.recIndicator);
				holder.recIndicator.setImageResource(R.drawable.ic_record);
				holder.title = (TextView) convertView.findViewById(R.id.title);
				holder.channelName = (TextView) convertView.findViewById(R.id.channelName);
				holder.date = (TextView) convertView.findViewById(R.id.date);
				convertView.setTag(holder);
			} else {
				holder = (ViewHolder) convertView.getTag();
			}
			Timer o = getItem(position);
			if (o != null) {
				holder.title.setText(o.getTitle());
				holder.channelName.setText(o.getChannelName());
				String date = DateUtils.getDateInLocalFormat(o.getStart());
				if (DateUtils.isToday(o.getStart().getTime())) {
					date = getResources().getString(R.string.today);
				} else if (DateUtils.isTomorrow(o.getStart().getTime())) {
					date = getResources().getString(R.string.tomorrow);
				}
				holder.layout.setError(o.isFlagSet(Timer.FLAG_EXECUTABLE));
				holder.layout.setDisabled(o.isFlagSet(Timer.FLAG_DISABLED));
				String start = DateUtils.getTimeInLocalFormat(getActivity(), o.getStart());
				String end = DateUtils.getTimeInLocalFormat(getActivity(), o.getEnd());
				holder.date.setText(date + "  " + start + " - " + end);
				holder.recIndicator.setVisibility(o.isFlagSet(Timer.FLAG_RECORDING) ? View.VISIBLE : View.GONE);
			}

			return convertView;
		}
	}

	/* (non-Javadoc)
     * @see org.dvbviewer.controller.ui.base.BaseListFragment#onListItemClick(android.widget.ListView, android.view.View, int, long)
     */
	@Override
	public void onListItemClick(ListView parent, View view, int position, long id) {
		if (actionMode) {
			view.setSelected(!view.isSelected());
			int count = getCheckedItemCount();
			updateActionModeTitle(count);
			if (getCheckedItemCount() == 0) {
				mode.finish();
			}
		} else {
			if (UIUtils.isTablet(getActivity())) {
				onDestroyActionMode(mode);
				Timer timer = mAdapter.getItem(position);
				TimerDetails timerdetails = TimerDetails.newInstance();
				Bundle args = TimerDetails.buildBundle(timer);
				timerdetails.setArguments(args);
				timerdetails.setTargetFragment(this, 0);
				timerdetails.setOnTimerEditedListener(this);
				timerdetails.show(getActivity().getSupportFragmentManager(), TimerDetails.class.getName());
			} else {
				getListView().setItemChecked(position, !getListView().isItemChecked(position));
				Timer timer = mAdapter.getItem(position);
				Intent timerIntent = new Intent(getActivity(), TimerDetailsActivity.class);
				Bundle extras = TimerDetails.buildBundle(timer);
				timerIntent.putExtras(extras);
				startActivityForResult(timerIntent, TimerDetails.TIMER_RESULT);
			}
		}

	}

	/* (non-Javadoc)
 * @see android.support.v4.app.Fragment#onSaveInstanceState(android.os.Bundle)
 */
	@Override
	public void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putBoolean(ACTION_MODE, actionMode);
		outState.putInt(CHECKED_ITEM_COUNT, getCheckedItemCount());
	}

	/* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onActivityResult(int, int, android.content.Intent)
	 */
	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		if (requestCode == TimerDetails.TIMER_RESULT && resultCode == TimerDetails.RESULT_CHANGED) {
            timerSavedAction();
		}
	}

    private void timerSavedAction() {
        getView().post(new Runnable() {
            @Override
            public void run() {
                Snackbar snackbar = Snackbar
                        .make(getListView(), R.string.timer_saved, Snackbar.LENGTH_LONG);
                snackbar.show();
                refresh();
            }
        });
    }

    /* (non-Javadoc)
     * @see com.actionbarsherlock.app.SherlockFragment#onCreateOptionsMenu(android.view.Menu, android.view.MenuInflater)
     */
	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
		inflater.inflate(R.menu.timer_list, menu);
	}

	/* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onCreateActionMode(com.actionbarsherlock.view.ActionMode, com.actionbarsherlock.view.Menu)
	 */
	@Override
	public boolean onCreateActionMode(ActionMode mode, Menu menu) {
		actionMode = true;
		getActivity().getMenuInflater().inflate(R.menu.actionmode_recording, menu);
		return true;
	}

	/* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onPrepareActionMode(com.actionbarsherlock.view.ActionMode, com.actionbarsherlock.view.Menu)
	 */
	@Override
	public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
		return false;
	}

	/* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onActionItemClicked(com.actionbarsherlock.view.ActionMode, com.actionbarsherlock.view.MenuItem)
	 */
	@Override
	public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
		switch (item.getItemId()) {
		case R.id.menuDelete:
			/**
			 * Alertdialog to confirm the delete of Recordings
			 */
			AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
			builder.setMessage(getResources().getString(R.string.confirmDelete)).setPositiveButton(getResources().getString(R.string.yes), this).setNegativeButton(getResources().getString(R.string.no), this).show();
			break;

		default:
			break;
		}
		return true;
	}

	/**
	 * The Class TimerDeleter.
	 *
	 * @author RayBa
	 * @date 07.04.2013
	 */
	public static class TimerDeleter extends AsyncTask<Timer, Void, Boolean> {

		AsyncCallback	callback;

		/**
		 * Instantiates a new timer deleter.
		 *
		 * @param callback the callback
		 * @author RayBa
		 * @date 07.04.2013
		 */
		public TimerDeleter(AsyncCallback callback) {
			this.callback = callback;
		}

		/* (non-Javadoc)
		 * @see android.os.AsyncTask#onPreExecute()
		 */
		@Override
		protected void onPreExecute() {
			super.onPreExecute();
			callback.onAsyncActionStart();
		}

		/* (non-Javadoc)
		 * @see android.os.AsyncTask#onPostExecute(java.lang.Object)
		 */
		@Override
		protected void onPostExecute(Boolean result) {
			super.onPostExecute(result);
			if (result) {
				callback.onAsyncActionStop();
			}

		}

		/* (non-Javadoc)
		 * @see android.os.AsyncTask#doInBackground(Params[])
		 */
		@Override
		protected Boolean doInBackground(Timer... params) {
			int count = params.length;
			if (count <= 0) {
				return false;
			}
			for (int i = 0; i < count; i++) {
				try {
					Thread.sleep(1000);
				} catch (InterruptedException e1) {
					// TODO Auto-generated catch block
					e1.printStackTrace();
				}
				try {
					ServerRequest.getRSString(ServerConsts.REC_SERVICE_URL + ServerConsts.URL_TIMER_DELETE + params[i].getId());
				} catch (AuthenticationException e) {
					e.printStackTrace();
				} catch (DefaultHttpException e) {
					e.printStackTrace();
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
			return true;
		}

	}

	/* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onDestroyActionMode(com.actionbarsherlock.view.ActionMode)
	 */
	@Override
	public void onDestroyActionMode(ActionMode mode) {
		actionMode = false;
		clearSelection();
	}

	/**
	 * Clear selection.
	 *
	 * @author RayBa
	 * @date 07.04.2013
	 */
	private void clearSelection() {
		for (int i = 0; i < getListAdapter().getCount(); i++) {
			getListView().setItemChecked(i, false);
		}
	}

	/* (non-Javadoc)
	 * @see android.content.DialogInterface.OnClickListener#onClick(android.content.DialogInterface, int)
	 */
	@Override
	public void onClick(DialogInterface dialog, int which) {
		switch (which) {
		case DialogInterface.BUTTON_POSITIVE:
			SparseBooleanArray checkedPositions = getListView().getCheckedItemPositions();
			if (checkedPositions != null && checkedPositions.size() > 0) {
				int size = checkedPositions.size();
				TimerDeleter deleter = new TimerDeleter(TimerList.this);
				List<Timer> timers = new ArrayList<Timer>();
				for (int i = 0; i < size; i++) {
					if (checkedPositions.valueAt(i)) {
						timers.add(mAdapter.getItem(checkedPositions.keyAt(i)));
					}
				}
				Timer[] array = new Timer[timers.size()];
				deleter.execute(timers.toArray(array));
			}
			mode.finish();
			break;

		case DialogInterface.BUTTON_NEGATIVE:
			// No button clicked
			break;
		}
	}

	@Override
	public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
		getListView().setItemChecked(position, true);
		int count = getCheckedItemCount();
		if (actionMode == false) {
			actionMode = true;
			AppCompatActivity activty = (AppCompatActivity) getActivity();
			mode = activty.startSupportActionMode(TimerList.this);
		}
		updateActionModeTitle(count);
		return true;
	}

	private void updateActionModeTitle(int count) {
		mode.setTitle(count + " " + getResources().getString(R.string.selected));
	}

	/**
	 * Refresh.
	 *
	 * @author RayBa
	 * @date 07.04.2013
	 */
	private void refresh() {
		getLoaderManager().restartLoader(0, getArguments(), this);
		setListShown(false);
	}

	/* (non-Javadoc)
	 * @see org.dvbviewer.controller.ui.base.BaseActivity.AsyncCallback#onAsyncActionStart()
	 */
	@Override
	public void onAsyncActionStart() {
		progressDialog = new ProgressDialog(getActivity());
		progressDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
		progressDialog.setMessage(getResources().getString(R.string.busyDeleteTimer));
		progressDialog.setIndeterminate(true);
		progressDialog.setCancelable(false);
		progressDialog.show();
	}

	/* (non-Javadoc)
	 * @see org.dvbviewer.controller.ui.base.BaseActivity.AsyncCallback#onAsyncActionStop()
	 */
	@Override
	public void onAsyncActionStop() {
		progressDialog.dismiss();
		refresh();
	}
	
	/* (non-Javadoc)
	 * @see com.actionbarsherlock.app.SherlockFragment#onOptionsItemSelected(android.view.MenuItem)
	 */
	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		int itemId = item.getItemId();
		switch (itemId) {
		case R.id.menuRefresh:
			refresh();
			return true;

		default:
			return false;
		}
	}
}

File: dvbViewerController/src/main/java/org/dvbviewer/controller/ui/fragments/TimerList.kt
/*
 * Copyright © 2013 dvbviewer-controller Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.dvbviewer.controller.ui.fragments

import android.content.Context
import android.content.DialogInterface
import android.content.DialogInterface.OnClickListener
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.AdapterView
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.appcompat.view.ActionMode.Callback
import androidx.loader.app.LoaderManager.LoaderCallbacks
import androidx.loader.content.Loader
import okhttp3.ResponseBody
import org.dvbviewer.controller.R
import org.dvbviewer.controller.data.version.TimerRepository
import org.dvbviewer.controller.entities.Timer
import org.dvbviewer.controller.io.api.APIClient
import org.dvbviewer.controller.io.api.DMSInterface
import org.dvbviewer.controller.ui.base.AsyncLoader
import org.dvbviewer.controller.ui.base.BaseListFragment
import org.dvbviewer.controller.ui.phone.TimerDetailsActivity
import org.dvbviewer.controller.ui.widget.ClickableRelativeLayout
import org.dvbviewer.controller.utils.*
import retrofit2.Call
import retrofit2.Response
import java.util.*


/**
 * The Class TimerList.
 *
 * @author RayBa
 * @date 07.04.2013
 */
class TimerList : BaseListFragment(), LoaderCallbacks<List<Timer>>, Callback, OnClickListener, TimerDetails.OnTimerEditedListener, AdapterView.OnItemLongClickListener {

    private lateinit var mAdapter: TimerAdapter
    private var mode: ActionMode? = null
    private var actionMode: Boolean = false
    private lateinit var repository: TimerRepository

    /* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onCreate(android.os.Bundle)
	 */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mAdapter = TimerAdapter(context)
        setHasOptionsMenu(true)
        val dmsInterface = APIClient.client.create(DMSInterface::class.java)
        repository = TimerRepository(dmsInterface)

    }

    /* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onActivityCreated(android.os.Bundle)
	 */
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        listAdapter = mAdapter
        val loader = loaderManager.initLoader(0, savedInstanceState, this)
        setListShown(!(!isResumed || loader.isStarted))
        listView!!.choiceMode = ListView.CHOICE_MODE_MULTIPLE
        listView!!.onItemLongClickListener = this
        setEmptyText(resources.getString(R.string.no_timer))
        if (savedInstanceState != null && savedInstanceState.getBoolean(ACTION_MODE, false)) {
            val activity = activity as AppCompatActivity?
            mode = activity!!.startSupportActionMode(this)
            updateActionModeTitle(savedInstanceState.getInt(CHECKED_ITEM_COUNT))
        } else {
            activity!!.setTitle(R.string.timer)
        }
    }

    /* (non-Javadoc)
	 * @see android.support.v4.app.LoaderManager.LoaderCallbacks#onCreateLoader(int, android.os.Bundle)
	 */
    override fun onCreateLoader(arg0: Int, arg1: Bundle?): Loader<List<Timer>> {
        return object : AsyncLoader<List<Timer>>(context!!) {

            override fun loadInBackground(): List<Timer> {
                try {
                    return repository.getTimer()!!
                } catch (e: Exception) {
                    catchException(javaClass.simpleName, e)
                }
                return Collections.emptyList()
            }
        }
    }

    /* (non-Javadoc)
	 * @see android.support.v4.app.LoaderManager.LoaderCallbacks#onLoadFinished(android.support.v4.content.Loader, java.lang.Object)
	 */
    override fun onLoadFinished(arg0: Loader<List<Timer>>, arg1: List<Timer>) {
        mAdapter.items = arg1
        setListShown(true)
    }

    /* (non-Javadoc)
	 * @see android.support.v4.app.LoaderManager.LoaderCallbacks#onLoaderReset(android.support.v4.content.Loader)
	 */
    override fun onLoaderReset(arg0: Loader<List<Timer>>) {
        arg0.reset()
        if (isVisible) {
            setListShown(true)
        }
    }

    override fun timerEdited(timer: Timer?) {
        timer?.let {
            setListShown(false)
            repository.saveTimer(it).enqueue(object : retrofit2.Callback<ResponseBody> {
                override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                    refresh()
                    sendMessage(R.string.timer_saved)
                    logEvent(EVENT_TIMER_EDITED)
                }

                override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                    setListShown(true)
                    sendMessage(R.string.error_common)
                }
            })
        }
    }

    /**
     * The Class ViewHolder.
     *
     * @author RayBa
     * @date 07.04.2013
     */
    private class ViewHolder {
        internal var layout: ClickableRelativeLayout? = null
        internal var recIndicator: ImageView? = null
        internal var title: TextView? = null
        internal var channelName: TextView? = null
        internal var date: TextView? = null
    }

    /**
     * The Class TimerAdapter.
     *
     * @author RayBa
     * @date 07.04.2013
     */
    inner class TimerAdapter
    /**
     * The Constructor.
     *
     * @param context the context
     * @author RayBa
     * @date 04.06.2010
     * @description Instantiates a new recording adapter.
     */
    (context: Context?) : ArrayListAdapter<Timer>() {

        /*
		 * (non-Javadoc)
		 *
		 * @see android.widget.ArrayAdapter#getView(int, android.view.View,
		 * android.view.ViewGroup)
		 */
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var convertView = convertView
            val holder: ViewHolder
            if (convertView == null) {
                convertView = LayoutInflater.from(context).inflate(R.layout.list_item_timer, parent, false)
                holder = ViewHolder()
                holder.layout = convertView as ClickableRelativeLayout?
                holder.recIndicator = convertView!!.findViewById(R.id.recIndicator)
                holder.title = convertView.findViewById(R.id.title)
                holder.channelName = convertView.findViewById(R.id.channelName)
                holder.date = convertView.findViewById(R.id.date)
                convertView.tag = holder
            } else {
                holder = convertView.tag as ViewHolder
            }
            val o = getItem(position)
            if (o != null) {
                holder.title!!.text = o.title
                holder.channelName!!.text = o.channelName
                var date = DateUtils.getDateInLocalFormat(o.start)
                if (DateUtils.isToday(o.start!!.time)) {
                    date = resources.getString(R.string.today)
                } else if (DateUtils.isTomorrow(o.start!!.time)) {
                    date = resources.getString(R.string.tomorrow)
                }
                holder.layout!!.isError = o.isFlagSet(Timer.FLAG_EXECUTABLE)
                holder.layout!!.isDisabled = o.isFlagSet(Timer.FLAG_DISABLED)
                val start = DateUtils.getTimeInLocalFormat(context, o.start)
                val end = DateUtils.getTimeInLocalFormat(context, o.end)
                holder.date!!.text = "$date  $start - $end"
                holder.recIndicator!!.visibility = if (o.isFlagSet(Timer.FLAG_RECORDING)) View.VISIBLE else View.GONE
            }

            return convertView!!
        }
    }

    /* (non-Javadoc)
     * @see org.dvbviewer.controller.ui.base.BaseListFragment#onListItemClick(android.widget.ListView, android.view.View, int, long)
     */
    override fun onListItemClick(parent: ListView, view: View, position: Int, id: Long) {
        if (actionMode) {
            view.isSelected = !view.isSelected
            val count = checkedItemCount
            updateActionModeTitle(count)
            if (checkedItemCount == 0) {
                mode!!.finish()
            }
        } else {
            if (UIUtils.isTablet(context!!)) {
                onDestroyActionMode(mode)
                val timer = mAdapter.getItem(position)
                val timerdetails = TimerDetails.newInstance()
                val args = TimerDetails.buildBundle(timer)
                timerdetails.arguments = args
                timerdetails.setTargetFragment(this, 0)
                timerdetails.setOnTimerEditedListener(this)
                timerdetails.show(activity!!.supportFragmentManager, TimerDetails::class.java.name)
            } else {
                listView!!.setItemChecked(position, !listView!!.isItemChecked(position))
                val timer = mAdapter.getItem(position)
                val timerIntent = Intent(context, TimerDetailsActivity::class.java)
                val extras = TimerDetails.buildBundle(timer)
                timerIntent.putExtras(extras)
                startActivityForResult(timerIntent, TimerDetails.TIMER_RESULT)
            }
        }

    }

    /* (non-Javadoc)
 * @see android.support.v4.app.Fragment#onSaveInstanceState(android.os.Bundle)
 */
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(ACTION_MODE, actionMode)
        outState.putInt(CHECKED_ITEM_COUNT, checkedItemCount)
    }

    /* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onActivityResult(int, int, android.content.Intent)
	 */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == TimerDetails.RESULT_CHANGED) {
            val timer: Timer? = data?.getSerializableExtra("timer") as Timer?
            timerEdited(timer)
        }
    }

    /* (non-Javadoc)
     * @see com.actionbarsherlock.app.SherlockFragment#onCreateOptionsMenu(android.view.Menu, android.view.MenuInflater)
     */
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.timer_list, menu)
    }

    /* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onCreateActionMode(com.actionbarsherlock.view.ActionMode, com.actionbarsherlock.view.Menu)
	 */
    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        actionMode = true
        activity!!.menuInflater.inflate(R.menu.actionmode_recording, menu)
        return true
    }

    /* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onPrepareActionMode(com.actionbarsherlock.view.ActionMode, com.actionbarsherlock.view.Menu)
	 */
    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        return false
    }

    /* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onActionItemClicked(com.actionbarsherlock.view.ActionMode, com.actionbarsherlock.view.MenuItem)
	 */
    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menuDelete -> {
                /**
                 * Alertdialog to confirm the delete of Recordings
                 */
                val builder = AlertDialog.Builder(context!!)
                builder.setMessage(resources.getString(R.string.confirmDelete)).setPositiveButton(resources.getString(R.string.yes), this).setNegativeButton(resources.getString(R.string.no), this).show()
            }

            else -> {
            }
        }
        return true
    }

    /**
     * The Class TimerDeleter.
     *
     * @author RayBa
     * @date 07.04.2013
     */

    /* (non-Javadoc)
	 * @see com.actionbarsherlock.view.ActionMode.Callback#onDestroyActionMode(com.actionbarsherlock.view.ActionMode)
	 */
    override fun onDestroyActionMode(mode: ActionMode?) {
        actionMode = false
        clearSelection()
    }

    /**
     * Clear selection.
     *
     * @author RayBa
     * @date 07.04.2013
     */
    private fun clearSelection() {
        for (i in 0 until listAdapter!!.count) {
            listView!!.setItemChecked(i, false)
        }
    }

    /* (non-Javadoc)
	 * @see android.content.DialogInterface.OnClickListener#onClick(android.content.DialogInterface, int)
	 */
    override fun onClick(dialog: DialogInterface, which: Int) {
        when (which) {
            DialogInterface.BUTTON_POSITIVE -> {
                val checkedPositions = listView!!.checkedItemPositions
                if (checkedPositions != null && checkedPositions.size() > 0) {
                    setListShown(false)
                    val size = checkedPositions.size()
                    val timers = ArrayList<Timer>()
                    for (i in 0 until size) {
                        if (checkedPositions.valueAt(i)) {
                            timers.add(mAdapter.getItem(checkedPositions.keyAt(i)))
                        }
                    }
                    repository.deleteTimer(timers).enqueue(object : retrofit2.Callback<ResponseBody> {
                        override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                            refresh()
                            sendMessage(R.string.timer_deleted)
                            logEvent(EVENT_TIMER_DELETED)
                        }

                        override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                            setListShown(true)
                            sendMessage(R.string.error_common)
                        }
                    })
                }
                mode?.finish()
            }

            DialogInterface.BUTTON_NEGATIVE -> {
            }
        }// No button clicked
    }

    override fun onItemLongClick(parent: AdapterView<*>, view: View, position: Int, id: Long): Boolean {
        listView!!.setItemChecked(position, true)
        val count = checkedItemCount
        if (actionMode == false) {
            actionMode = true
            val activty = activity as AppCompatActivity?
            mode = activty?.startSupportActionMode(this@TimerList)
        }
        updateActionModeTitle(count)
        return true
    }

    private fun updateActionModeTitle(count: Int) {
        mode?.title = count.toString() + " " + resources.getString(R.string.selected)
    }

    /**
     * Refresh.
     *
     * @author RayBa
     * @date 07.04.2013
     */
    private fun refresh() {
        loaderManager.restartLoader(0, arguments, this)
        setListShown(false)
    }

    /* (non-Javadoc)
	 * @see com.actionbarsherlock.app.SherlockFragment#onOptionsItemSelected(android.view.MenuItem)
	 */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        when (itemId) {
            R.id.menuRefresh -> {
                refresh()
                return true
            }

            else -> return false
        }
    }

    companion object {

        val ACTION_MODE = "action_mode"
        val CHECKED_ITEM_COUNT = "checked_item_count"
    }
}
